README.md
# OwlVoice
Chrome Web Extension for all those who are tired of password managers and want to carry their database around with them or store it on a cloud storage.

## How it works
Take the `domain + Username + Globalsecret + CustomSecret`make an SHA2 hash and encode the complete hash to base85


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.2.0
### Changed
- Permissions

## 1.1.0
### Added
- .gitignore
- Changelog

## 1.0.0
### Added
- Initial commit


background.js
importScripts('/lib/base85encoder.js');

// global identifier
const CONTEXT_MENU_ID = "OwlVoice";

// In-page cache of the user's options
let options = {};

/**
 * Loads Storage Data
 */
chrome.storage.sync.get({
    globalSecret: Math.random().toString(36).substr(2),
    hashVariant: 'SHA-256',
    passwordLength: 12
}, items => {
    options = items;
});

/**
 * On Install set defined globalSecret
 */
chrome.runtime.onInstalled.addListener(() => {
    chrome.storage.sync.set({
        globalSecret: options.globalSecret
    });

});

/**
 * Opens permanent connection to communicate with popup
 */
chrome.runtime.onConnect.addListener(function(port) {
    port.onMessage.addListener(async function(msg) {
        const password = await getPassword(msg.content);
        port.postMessage({type: msg.type, password: password});
    });
});

/**
 * Adds context menu entry
 */
chrome.contextMenus.create({
    title: "Get Password for: %s",
    id: CONTEXT_MENU_ID,
    contexts: ["selection"]
});

/**
 * Set an EventListener to context menu entry
 */
chrome.contextMenus.onClicked.addListener(contextFunction);

async function getPassword(content) {
    const hash = await digest(content, options.hashVariant);
    const base85 = new Base85Encoder();
    const password = base85.encode(hash);

    return password.substr(0, options.passwordLength);
}

/**
 * Hash given string with given digest
 *
 * @param {String} value Value
 * @param {String} type  Digest type
 *
 * @return {Promise<string>}
 */
async function digest(value, type = 'SHA-256') {

    return Array.from(
        new Uint8Array(await crypto.subtle.digest(type, new TextEncoder().encode(value)))
    ).map(b => b.toString(16).padStart(2, '0')).join('');

}

/**
 * Creates Password from selected String and send it to password field

 * @param {Object} info
 * @param {Object} tab
 * @returns {Promise<void>}
 */
async function contextFunction(info, tab) {
    if (info.menuItemId !== CONTEXT_MENU_ID) {
        return;
    }

    const urlObject = new URL(tab.url);
    const domain = urlObject.protocol + '//' + urlObject.hostname;

    const content = domain + options.globalSecret + info.selectionText.trim();
    const password = await getPassword(content);

    chrome.tabs.sendMessage(tab.id, {'password': password});
}

options/options.js
const hashElement = document.getElementById('hashtype');
const globalSecretElement = document.getElementById('globalsecret');
const passwordLengthElement = document.getElementById('passwordlength');
const passwordLengthValueElement = document.getElementById('passwordlengthvalue');

/**
 *
 */
function saveOptions() {
    let globalSecret = globalSecretElement.value;
    let hashVariant = hashElement.value;
    let passwordLength = passwordLengthElement.value;

    chrome.storage.sync.set({
        globalSecret: globalSecret,
        hashVariant: hashVariant,
        passwordLength: passwordLength
    }, function() {
        let status = document.getElementById('status');
        status.textContent = 'Options saved.';

        setTimeout(function() {
            status.textContent = '';
        }, 750);
    });
}

/**
 *
 */
function restoreOptions() {
    chrome.storage.sync.get({
        globalSecret: Math.random().toString(36).substr(2),
        hashVariant: 'SHA-256',
        passwordLength: 12
    }, function(items) {
        globalSecretElement.value = items.globalSecret || Math.random().toString(36).substr(2);
        hashElement.value = items.hashVariant;
        passwordLengthElement.value = items.passwordLength;
        passwordLengthValueElement.value = items.passwordLength;
    });
}

/**
 *
 */
document.addEventListener('DOMContentLoaded', restoreOptions);
document.getElementById('save').addEventListener('click', saveOptions);
passwordLengthElement.addEventListener('input', () => {
    passwordLengthValueElement.value = passwordLengthElement.value;
});

lib/Base85Encoder.js
/**
 * base85 Encoder
 *
 * How it works:
 * 1. Divide a text into 4-character chunks
 * 2. Convert the chunks of text to binary
 * 3. Padding binary data to 32 bit
 * 4. Calculate based on base85
 * 5. Convert the numbers to characters
 * 6. Output the concatenated characters
 */
class Base85Encoder {

    constructor(useCustomAlphabet = false) {
        this.useCustomAlphabet = useCustomAlphabet;
        this.customAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-:+=^!/*?&<>()[]{}@%$#';
    }

    /**
     * Encode given String
     *
     * @param {string} text
     * @returns {string}
     */
    encode(text) {
        let blocks = [];
        let output = [];

        const textChunks = this.chunkString(text);

        textChunks.forEach(chunk => {
            let binary = this.textToBinary(chunk).join('');
            blocks.push(this.padBinary(binary));
        });

        blocks.forEach(block => {
            let separatedValues = this.loopedDivision(parseInt(block, 2));
            output = [].concat(output, this.lookupChar(separatedValues));
        });

        return output.join('');
    }

    /**
     * Calculate to base 85
     *
     * divide by 85^4 & modulo 85^4 as new start value
     * divide by 85^3 & modulo 85^3 as new start value
     * divide by 85^2 & modulo 85^2 as new start value
     * divide by 85   & modulo 85   as new start value
     *
     * @param {Number} value
     * @returns {Number[]}
     */
    loopedDivision(value) {
        let values = [];
        let modulo = value;

        for (let i = 4, len = 1; i >= len; i--) {

            let divisor = Math.pow(85, i);
            let calculatedValue = ~~(modulo / divisor);
            modulo = modulo % divisor;

            values.push(calculatedValue);
        }

        values.push(modulo);
        return values;
    }

    /**
     * Transform number to character
     * @param {Number[]} value
     * @returns {string[]}
     */
    lookupChar(value) {
        let values = [];

        value.forEach(item => {

            if (!this.useCustomAlphabet) {
                item += 33;
            }

            values.push(String.fromCharCode(item));

        });

        return values;
    }

    /**
     * Split given string to chunks
     *
     * @param {string} text
     * @returns {boolean|string[]}
     */
    chunkString(text = '') {
        if (text.length <= 0) {
            return false;
        }

        return text.match(/.{1,4}/g);
    }

    /**
     * Padding the binary data to 32-bit
     *
     * @param binary
     * @returns {boolean|string}
     */
    padBinary(binary = '') {
        if (binary.length <= 0) {
            return false;
        }

        return binary.toString().padEnd(32, '0');
    }

    /**
     * Transform string to binary
     *
     * @param text
     * @returns {boolean|string[]}
     */
    textToBinary(text='') {
        if (text.length <= 0) {
            return false;
        }

        return text.split('').map(char => {
            return char.charCodeAt(0).toString(2).padStart(8, '0');
        });
    }
}



popup/popup.js
const showDomainElement = document.getElementById('domain');
const usernameElement = document.getElementById('username');
const customSecretElement = document.getElementById('secret');
const feedbackElement = document.getElementById('feedback');

const copyButton = document.getElementById('copy');
const autofillButton = document.getElementById('fill');
const clearLink = document.getElementById('clear');

const port = chrome.runtime.connect({
    name: "Owl Voice"
});

let domain = 'localhost';
let options = {}
let currentTabId;

/**
 * Load storage data
 */
chrome.storage.sync.get({
    globalSecret: Math.random().toString(36).substr(2),
    hashVariant: 'SHA-256',
    passwordLength: 12
}, items => {
    options = items;
});

/**
 * Get active Tab and save URL and TabID
 */
chrome.tabs.query({ active: true, currentWindow: true }, tabs => {
    const urlObject = new URL(tabs[0].url);
    domain = urlObject.protocol + '//' + urlObject.hostname;
    showDomainElement.innerText = domain;
    currentTabId = tabs[0].id;
});

/**
 * Get Password from the background.js
 *
 * @param {string} content
 * @param {string} type
 * @returns {void}
 */
async function getPassword(content, type) {

    port.postMessage({type: type, content: content});

    port.onMessage.addListener(async (message) => {
        if (message.type === 'clipboard') {
            navigator.clipboard.writeText(message.password).then(() =>{
                feedbackElement.innerText = 'Saved to clipboard'
                clearFeedback();
            });
        } else if (message.type === 'autofill') {
            chrome.tabs.sendMessage(currentTabId, {'password': message.password});
            feedbackElement.innerText = 'Password transmitted';
            clearFeedback();
        }
    });
}

/**
 * Build content string and get Password from the background.js
 * @param {String} username
 * @param {String} customSecret
 * @param {String} type
 * @returns {Promise<string>}
 */
async function generatePassword(username, customSecret, type) {
    const content = domain + options.globalSecret + username + customSecret;
    await getPassword(content, type);
}

function clearFeedback() {
    setTimeout(() => {
        feedbackElement.innerText = '';
    }, 4000)
}

/**
 * Copy button action
 * Generates password an save it to clipboard
 */
copyButton.addEventListener('click', async () => {
    const username = usernameElement.value;
    const customSecret = customSecretElement.value;

    await generatePassword(username, customSecret, 'clipboard');
});

/**
 * Autofill button action
 * Generates Password and autofill password field
 */
autofillButton.addEventListener('click', async () => {
    const username = usernameElement.value;
    const customSecret = customSecretElement.value;

    await generatePassword(username, customSecret, 'autofill');
});

/**
 * Reset Input Fields and clear clipboard
 */
clearLink.addEventListener('click', () => {
    usernameElement.value = '';
    customSecretElement.value = '';

    navigator.clipboard.writeText('').then(() => {feedbackElement.innerText = 'Clipboard cleared'});
    clearFeedback();
});